### PR TITLE
[no_early_kickoff] [Tune] Update `Tuner.restore` usage to prepare for `trainable` becoming a required arg

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

This PR updates `Tuner.restore` usage in tests and outward-facing docs to include the trainable. This arg will be required starting in 2.5. This PR also fixes some links in error messages.

There are a few usage examples that haven't been updated -- this is because https://github.com/ray-project/ray/pull/33495 is already handling those.

## Related issue number

https://github.com/ray-project/ray/issues/33546

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by richardliaw — APPROVED** [review]
whats the reason why there is such a big diff?

**Comment by justinvyu** [comment]
@richardliaw I had a merge conflict on a Jupyter notebook, and I might have changed a bunch of metadata in the cells. I can redo this to make the diff small if needed.

File: python/ray/tune/impl/tuner_internal.py [hunk 1]
```diff
@@ -260,8 +260,8 @@ def _validate_overwrite_trainable(
                 "trainable_with_params = tune.with_parameters(trainable, ...)\n"
                 "tuner = tune.Tuner.restore(\n"
                 "    ..., trainable=trainable_with_params\n"
-                ")\n\nSee https://docs.ray.io/en/master/tune/api_docs/trainable.html"
-                "#tune-with-parameters for more details."
+                ")\n\nSee https://docs.ray.io/en/latest/tune/api/doc/"
+                "ray.tune.with_parameters.html for more details."
```
Comment (matthewdeng) [inline python/ray/tune/impl/tuner_internal.py:264]: Can you do something like:
```
:func:`ray.tune.with_parameters`
```

Comment (matthewdeng) [reply python/ray/tune/impl/tuner_internal.py:264]: Oh nevermind I see that this is an error message.

**Review by Yard1 — APPROVED** [review]

**Comment by Yard1** [comment]
@justinvyu can you resolve conflicts? we can merge it then!

**Comment by justinvyu** [comment]
The link check got fixed by another PR already. I changed a few other links to relative references (ex: in error messages), so we can still merge this one.

**Review by Yard1 — APPROVED** [review]

**Review by matthewdeng — APPROVED** [review]

**Review by krfricke — APPROVED** [review]
LGTM - just for reference, is there any documentation on the discussion/decision on providing a trainable? Will we remove trainable.pkl from the experiment dir?

**Comment by justinvyu** [comment]
@krfricke Here are some documents:

1. Tracker issue https://github.com/ray-project/ray/issues/33546
2. Launch doc: https://docs.google.com/document/d/1yQ3-nPPPK7bYxc0ubF5BtL4-XP0GxOWdlJuSzYk4-c4/edit#
3. API proposal: https://docs.google.com/document/d/180KDhwQ8EZH3FYW94nnN-uj8lsp2JX4VEFafpLNttIk/edit#

Summary of the decision:
- Users have the trainable on-hand when doing restoration.
- Tune shouldn't be responsible for serializing arbitrary user code in trainables (could be non-serializable, could capture object/actor refs in its scope).
- Users shouldn't need to worry about when to pass in the trainable again vs. when it's optional.
- Tuner.restore(..., trainable) and Tuner(trainable) are now both required args -- consistency between original and new run. `param_space` is optional for both.

**Pull request merged**